aten/src/ATen/native/Distributions.cpp

```diff
@@ -487,6 +487,21 @@ Tensor& multinomial_out(Tensor& result, const Tensor& self, int64_t n_sample, bo
   } else {
     result.resize_({n_sample});
   }
+
+  // Fast-path based on RobertoLat example.
+  // Reference:
+  // https://github.com/pytorch/pytorch/issues/11931#issuecomment-625882503
+  // Half is not supported on CPU.
+  if (!with_replacement &&
+      !(self.device().is_cpu() && self.scalar_type() == ScalarType::Half)) {
+    auto is_valid = ((self.max() < INFINITY) & (self.min() >= 0)).all().item();
+    TORCH_CHECK(is_valid.to<bool>(), "self contains either `inf`, `nan` or element < 0");
+    auto rand = at::empty_like(self).uniform_(0, 1, gen);
+    rand.log_().div_(self); //save memory with inplace operations
```

Comment:
uniform_ can produce 0. This by itself is not yet a problem, log will be `-inf`. However, if self happens to have 0 at the same place, it will become a problem, as here the value becomes `nan`, and will be selected by topk (I'm not sure what topk behavior is wrt nans, but I think it tries to propagate them).
Similarly, if `self` consists of 0-s only, the operation is not well defined. So maybe excluding zeros was not a horrible idea?

Reply:
I guess the `0` has to be valid prob as the below test depends on that assumption.
https://github.com/pytorch/pytorch/blob/2633a9cca1a2bb336919c9e2f3a8850852c9115a/test/test_cuda.py#L1651-L1683

Reply:
Ok, but then -inf/0 problem has to be addressed somehow, and probably distribution consisting of all zeros has to be invalid? (Currently we error out with " invalid multinomial distribution (sum of probabilities <= 0)", but I'm not sure where it happens).

Reply:
> and will be selected by topk (I'm not sure what topk behavior is wrt nans, but I think it tries to propagate them)

It does propagate `nan`.

```python
>>> torch.Tensor([0, 1, 3, float('nan')]).topk(2)
torch.return_types.topk(
values=tensor([nan, 3.]),
indices=tensor([3, 2]))
```

However, I tried the following

```python
>>> torch.Tensor([float('inf')]).div(0)
tensor([inf])
>>> torch.Tensor([float('-inf')]).div(0)
tensor([-inf])
>>> torch.Tensor([float('nan')]).div(0)
tensor([nan])
>>> torch.Tensor([float('0')]).div(0)
tensor([nan])
```

Looking at this `-inf/0` shouldn't be an issue as it will stay as `-inf`.

Will add a check for ` " invalid multinomial distribution (sum of probabilities <= 0)"`. I guess this will cover all bases.